clients/src/test/java/org/apache/kafka/clients/consumer/internals/FetcherTest.java

```diff
@@ -2940,6 +2940,62 @@ private void verifySessionPartitions() {
         assertEquals(0, future.get());
     }
 
+    @Test
+    public void testFetcherSessionEpochUpdate() throws Exception {
```

Comment:
I guess there was no choice but to carefully tailor this test in order to hit the bug. We have to do it, but the downside is that its scope is narrow and may be difficult to keep it relevant as the code evolves. Anyway, hopefully at some point we'll get the time to move all network IO to the background thread and then we can simplify a lot of this.